cmd/restic/cmd_mount.go

```diff
@@ -0,0 +1,301 @@
+		snapshot, err := restic.LoadSnapshot(sn.repo, id)
+		if err != nil {
+			return nil, err
+		}
+		sn.knownSnapshots[snapshot.Time.Format(time.RFC3339)] = snapshot
+		ret = append(ret, fuse.Dirent{
+			Inode: binary.BigEndian.Uint64(id[:8]),
+			Type:  fuse.DT_Dir,
+			Name:  snapshot.Time.Format(time.RFC3339),
+		})
+	}
+
+	return ret, nil
+}
+
+func (sn *snapshots) Lookup(ctx context.Context, name string) (fs.Node, error) {
+	if _, ok := sn.knownSnapshots[name]; !ok {
+		// At least this snapshot is not cached. We use this opportunity to
+		// load all missing snapshots
+		for id := range sn.repo.List(backend.Snapshot, ctx.Done()) {
+			snapshot, err := restic.LoadSnapshot(sn.repo, id)
+			if err != nil {
+				return nil, err
+			}
+			snapshotName := snapshot.Time.Format(time.RFC3339)
+			sn.knownSnapshots[snapshotName] = snapshot
+			break
+		}
+	}
+
+	snapshot := sn.knownSnapshots[name]
+	tree, err := restic.LoadTree(sn.repo, snapshot.Tree)
+	if err != nil {
+		return nil, err
+	}
+	return &dir{
+		repo: sn.repo,
+		tree: tree,
+	}, nil
+	return nil, fuse.ENOENT
+}
+
+// Statically ensure that *dir implement those interface
+var _ = fs.HandleReadDirAller(&dir{})
+var _ = fs.NodeStringLookuper(&dir{})
+
+type dir struct {
+	repo  *repository.Repository
+	tree  *restic.Tree
+	inode uint64
+}
+
+func (d *dir) Attr(ctx context.Context, a *fuse.Attr) error {
+	a.Inode = d.inode
+	a.Mode = os.ModeDir | 0555
+	return nil
+}
+
+func (d *dir) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
+	ret := make([]fuse.Dirent, 0, len(d.tree.Nodes))
+
+	for _, node := range d.tree.Nodes {
+		var typ fuse.DirentType
+		switch {
+		case node.Mode.IsDir():
+			typ = fuse.DT_Dir
+		case node.Mode.IsRegular():
+			typ = fuse.DT_File
+		}
+
+		ret = append(ret, fuse.Dirent{
+			Inode: node.Inode,
+			Type:  typ,
+			Name:  node.Name,
+		})
+	}
+
+	return ret, nil
+}
+
+func (d *dir) Lookup(ctx context.Context, name string) (fs.Node, error) {
+	for _, node := range d.tree.Nodes {
+		if node.Name == name {
+			switch {
+			case node.Mode.IsDir():
+				subtree, err := restic.LoadTree(d.repo, node.Subtree)
+				if err != nil {
+					return nil, err
+				}
+				return &dir{
+					repo:  d.repo,
+					tree:  subtree,
+					inode: binary.BigEndian.Uint64(node.Subtree[:8]),
+				}, nil
+			case node.Mode.IsRegular():
+				return makeFile(d.repo, node)
+			}
+		}
+	}
+
+	return nil, fuse.ENOENT
+}
+
+// Statically ensure that *file implements the given interface
+var _ = fs.HandleReader(&file{})
+
+type file struct {
+	repo *repository.Repository
+	node *restic.Node
+
+	sizes []uint32
+
+	// cleartext contents
+	clearContent [][]byte
+}
+
+func makeFile(repo *repository.Repository, node *restic.Node) (*file, error) {
+	sizes := make([]uint32, len(node.Content))
+	for i, bid := range node.Content {
+		_, _, _, length, err := repo.Index().Lookup(bid)
+		if err != nil {
+			return nil, err
+		}
+		sizes[i] = uint32(length) - crypto.Extension
```

Comment:
can you explain the `- crypto.Extension`? somehow I feel like this doesn't belong here (not in the sense that the size will be wrong but in the sense that this file here should not be concerned with any crypto things and this here is the wrong level of abstraction for this).